third_party/jitterentropy/jitterentropy-base-user.h

```diff
@@ -0,0 +1,249 @@
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE, ALL OF
+ * WHICH ARE HEREBY DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF THIS SOFTWARE, EVEN IF NOT ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _JITTERENTROPY_BASE_USER_H
+#define _JITTERENTROPY_BASE_USER_H
+
+/*
+ * Set the following defines as needed for your environment
+ */
+/* Compilation for libgcrypt */
+#ifndef LIBGCRYPT
+#undef LIBGCRYPT
+#endif
+
+/* Compilation for OpenSSL */
+#ifndef OPENSSL
+#undef OPENSSL
+#endif
+
+#include <time.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sched.h>
+
+/* Timer-less entropy source */
+#ifdef JENT_CONF_ENABLE_INTERNAL_TIMER
+#include <pthread.h>
+#endif /* JENT_CONF_ENABLE_INTERNAL_TIMER */
+
+#ifdef LIBGCRYPT
+#include <config.h>
+#include "g10lib.h"
+#endif
+
+#ifdef OPENSSL
+#include <openssl/crypto.h>
+#ifdef OPENSSL_FIPS
+#include <openssl/fips.h>
+#endif
+#endif
+
+#ifdef AWSLC_FIPS
+#include <openssl/crypto.h>
+#endif
+
+#ifdef __MACH__
+#include <assert.h>
+#include <CoreServices/CoreServices.h>
+#include <mach/mach.h>
+#include <mach/mach_time.h>
+#include <unistd.h>
+#endif
+
+#ifdef __x86_64__
+
+# define DECLARE_ARGS(val, low, high)    unsigned long low, high
+# define EAX_EDX_VAL(val, low, high)     ((low) | (high) << 32)
+# define EAX_EDX_RET(val, low, high)     "=a" (low), "=d" (high)
+
+static inline void jent_get_nstime(uint64_t *out)
+{
+	DECLARE_ARGS(val, low, high);
+	__asm__ volatile("rdtsc" : EAX_EDX_RET(val, low, high));
+	*out = EAX_EDX_VAL(val, low, high);
+}
+
+#else /* __x86_64__ */
+
+static inline void jent_get_nstime(uint64_t *out)
+{
+	/* OSX does not have clock_gettime -- taken from
+	 * http://developer.apple.com/library/mac/qa/qa1398/_index.html */
+# ifdef __MACH__
+	*out = mach_absolute_time();
+# elif _AIX
+	/* clock_gettime() on AIX returns a timer value that increments in
+	 * steps of 1000
+	 */
+	uint64_t tmp = 0;
+	timebasestruct_t aixtime;
+	read_real_time(&aixtime, TIMEBASE_SZ);
+	tmp = aixtime.tb_high;
+	tmp = tmp << 32;
+	tmp = tmp | aixtime.tb_low;
+	*out = tmp;
+# else /* __MACH__ */
+	/* we could use CLOCK_MONOTONIC(_RAW), but with CLOCK_REALTIME
+	 * we get some nice extra entropy once in a while from the NTP actions
+	 * that we want to use as well... though, we do not rely on that
+	 * extra little entropy */
+	uint64_t tmp = 0;
+	struct timespec time;
+	if (clock_gettime(CLOCK_REALTIME, &time) == 0)
```

Comment:
While this is okay, if we can find the ARM equivalent of `rdtsc` I'd prefer that.